pdu/UDH.go

```diff
@@ -45,37 +53,30 @@ func (u *UDH) MarshalBinary() (b []byte, err error) {
 // The function returns the number of bytes read from src, and any error
 // detected while unmarshalling.
 func (u *UDH) UnmarshalBinary(src []byte) (int, error) {
-	if len(src) < 1 {
-		return 0, fmt.Errorf("Decode error UDHL %d underflow", 0)
+	if length := len(src); length > 1 && length < int(src[0]+1) {
+		return 0, fmt.Errorf("Decode error UDHL underflow")
 	}
-
-	udhl := int(src[0])
-	udhl++ // so it includes itself
-	ri := 1
-	if len(src) < udhl {
-		return ri, fmt.Errorf("Decode error InfoElement %d underflow", ri)
-	}
-
-	ies := []InfoElement(nil)
-	for ri < udhl {
-		if udhl < ri+2 {
-			return ri, fmt.Errorf("Decode error InfoElement %d underflow", ri)
+	length := src[0]
+	payload := src[1 : length+1]
+	header := UDH{}
+	var start, end byte
+	for i := byte(0); i < length; {
+		blockSize := payload[i+1]
+		start = i + 2
+		end = start + blockSize
+		if end <= byte(len(payload)) {
```

Comment:
This is wrong:
1. Don't downcast the index to prevent wrong overflow checking.
For example:
length = 255, i = 250, blockSize = 50
end will be < 100 which is pretty wrong because of overflow.
2. Don't downcast len of payload. let define a variable for that `payloadLen := len(payload)`
3. it must be `if end > payloadLen { return error }`